euphonic/design/03_design_document_v0.3.md

```diff
@@ -0,0 +1,531 @@
+|Attribute|Type|
+|---------|----|
+|Cell vectors| Quantity wrapped (3,3) float numpy.ndarray|
+|Number of ions|int|
+|Ion coordinates (fractional)|(n_ions, 3) array of floats|
+|Ion types|(n_ions,) list of strings|
+|Ion masses| Quantity wrapped (n_ions,) float numpy.ndarray|
+|Cell vector unit|string|
+|Ion mass unit units|string|
+
+Methods required for this object are:
+
+|Method|Other Input required|Output|
+|------|--------------|------|
+|Calculate reciprocal lattice|-|Reciprocal lattice ((3,3) array of floats)|
+|Calculate cell volume|-|Cell volume (float)|
+|Serialise|Filename|File|
+
+### Force Constants
+
+Force constants can be Fourier interpolated to calculate frequencies and
+eigenvectors at any q-point, it is therefore Euphonic's main data object and can
+be used to create most others. Force constants are calculated by an external
+code (e.g. CASTEP, Phonopy) and read to create this object. Force constants are
+useless on their own and require extra structure data (read in along with the
+force constants). The attributes required for this object are:
+
+|Attribute|Type|
+|---------|----|
+|Structure| Structure object|
+|Force constants|(n_cells, 3\*n_ions, 3\*n_ions) array of floats|
+|Cell origins|(n_cells, 3) array of ints|
+|Supercell matrix|(3, 3) array of ints|
+|Born charges (optional)|(n_ions, 3, 3) array of floats|
+|Dielectric permittivity tensor (optional)|(3, 3) array of floats|
+|Force constants unit|string|
+|Born charges unit|string|
+|Dielectric permittivity unit|string|
+
+Methods required on this object are:
+
+|Method|Other Input required|Output|
+|------|--------------|------|
+|Calculate phonon frequencies/eigenvectors (see [algorithms - 1.](04_algorithms.md))|Q-points|Phonon Data|
+|Calculate powder structure factor|Energy bins, \|Q\|-bins, neutron scattering lengths, temperature, Debye-Waller exponent(optional)|2D Spectrum object|
+|Serialise|Filename|File|
+
+### Phonon Data (at q-points)
+
+This contains phonon frequencies/eigenvectors calculated at arbitrary q-points.
+For dielectric materials, at the gamma point the frequencies are different
+depending on the direction of approach (LO-TO splitting). This should be
+included on dispersion plots so there must be a mechanism to contain the
+duplicated frequencies/eigenvectors at gamma points. They are contained in
+separate arrays, with an extra indexing array to determine which q-point they
+belong to.
+
+|Attribute|Type|
+|---------|----|
+|Structure|Structure object|
+|Number of q-points|int|
+|Q-points|(n_qpts, 3) array of floats|
+|Q-point weights|(n_qpts,) array of floats|
+|Frequencies|(n_qpts, 3*n_ions) array of floats|
+|Eigenvectors|(n_qpts, 3*n_ions, n_ions, 3) array of complex floats|
+|Split frequencies| (n_gamma_points, 3*n_ions) array of floats|
+|Split eigenvectors|(n_gamma_points, 3*n_ions, n_ions, 3) array of complex floats|
+|Gamma point indices|(n_gamma_points,) array of ints| 
+|Energy unit|string|
```

Comment:
This is a good point. I initially wrote it like this because I felt like if you pass in a list of `N` q-points, and the object that is returned has a list of `N + 1` q-points due to splitting that might be confusing. However, I've been having second thoughts about it and agree that it is overly complicated. If you think that having a repeated value is better and is the more 'usual' thing to do, I'm happy to do that instead, I know this is what CASTEP does at least.

Reply:
I see, it is a little uncomfortable that the output data is not going to match the input sequence. Pushing some of the values into another attribute hides this problem rather than solves it. We should clearly establish that the usage pattern for this object is to work with the data it produces and not try to combine it with leftover input data.

Maybe drop a note in the document  here clarifying that n_qpts does not equal the number of _unique_ q-points.

Reply:
I'll add a note on this